src/types.rs

```diff
@@ -921,3 +930,81 @@ impl Debug for Log {
             .finish()
     }
 }
+
+/// An Ethereum Transaction Receipt
+#[derive(Clone, Default, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TransactionReceipt {
+    /// Transaction type,
+    ///     - Some(3) for EIP-4844 transaction,
+    ///     - Some(2) for EIP-1559 transaction,
+    ///     - Some(1) for AccessList transaction,
+    ///     - None for Legacy
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub transaction_type: Option<U256>,
```

Comment:
This should either use a tagged enum similar to the `SignedTransaction` type if there are fields that only are set for certain kinds of transactions, or an enum should be used and not a raw numerical value.

*Edit*: This appears to be the care, `blob_gas_price` is only present for `Type:: Eip4844`.

Reply:
Yea that makes sense. So the whole structure is an enum. Would it his also mean that we don't even need to include the type field on the receipt?

Reply:
> Would it his also mean that we don't even need to include the type field on the receipt?

No, it is the ["tag"](https://serde.rs/enum-representations.html#internally-tagged) of the enum.

Reply:
Right. I see that now. However it looks like it might be a bit extreme to introduce 4 different types since there seem to be only 2 different variants of the fields. Most of the optional stuff seems to come from the most recently introduced type. 

I don't mind doing all three just wondering if it's worth all the copy-paste for similar structures. 

What would you do here?

Reply:
Note that you can reuse the structure:

```
enum TransactionReceipt {
    Legacy(LegacyTransactionReceipt),
    Eip2930(LegacyTransactionReceipt),
    Eip1559(LegacyTransactionReceipt),
    Eip4844(Eip4844TransactionReceipt),
}
```

Reply:
That much I'm aware of, but then comes the naming. Do you think legacy looks right there?

Reply:
> That much I'm aware of

Sorry, I just understood that when you said "However it looks like it might be a bit extreme to introduce 4 different types since there seem to be only 2 different variants of the fields", you were suggesting not to reuse the data struct across different variants, hence my suggestion.

Do you know if post EIP-4844 if _all_ transaction receipts will include the new field? I wouldn't be surprised if it were the case (same thing happened post EIP-1559 with the `effective_gas_price` field being on all transaction receipts.

Another potentially better strategy would be to have:

```rust
struct TransactionReceipt {
    pub kind: TransactionReceiptKind,
    
    // other stuff...
    pub other_fields: OtherFieldTypes,
}

enum TransactionReceiptKind {
    Legacy,
    Eip2930,
    Eip1559,
    Eip4844 {
        blob_gas_price: U256,
    }
}
```

Reply:
Unfortunately I do not know how to make these structured enum variants serialize.

JsonError(Error("invalid type: unit variant, expected struct variant"))

This is the end of the road for me. Would love some tips.

Reply:
This is resolved.